client/core/trade.go

```diff
@@ -367,92 +382,98 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
 			},
 		}
 
-		// First check that this isn't a match on its own cancel order. I'm not crazy
-		// about this, but I am detecting this case right now based on the Address
-		// field being an empty string.
-		isCancel := t.cancel != nil && msgMatch.Address == ""
-		if isCancel {
-			match.setStatus(order.MatchComplete)
-			includesCancellation = true
-			log.Infof("maker notification for cancel order received for order %s. match id = %s", oid, msgMatch.MatchID)
-			t.cancel.matches.maker = msgMatch
+		var qty uint64
+		if t.Type() == order.MarketOrderType && !trade.Sell {
+			qty = calc.BaseToQuote(msgMatch.Rate, msgMatch.Quantity)
 		} else {
-			includesTrades = true
-			t.matches[match.id] = match
+			qty = msgMatch.Quantity
 		}
+		newFill += qty
+
+		log.Infof("Starting negotiation for match %v for order %v with feeRateSwap = %v, quantity = %v",
+			mid, oid, feeRateSwap, qty)
+		newTrackers = append(newTrackers, match)
+	}
+
+	// Now that each Match in msgMatches has been validated, store them in the
+	// trackedTrade and the DB, and update each order's status and fill amount.
+	// First record any cancel order Match and update order status.
+	if cancelMatch != nil {
+		t.cancel.matches.maker = cancelMatch // taker is stored via processCancelMatch before negotiate
+		// Set the order status for both orders.
+		t.metaData.Status = order.OrderStatusCanceled
+		t.db.UpdateOrderStatus(t.cancel.ID(), order.OrderStatusExecuted)
+	}
+	for _, match := range newTrackers {
+		t.matches[match.id] = match
```

Comment:
In line with the any bad data = all bad data philosophy, the db and matches map are not updated until after each msgMatch is verified.